narwhals/_spark_like/utils.py

```diff
@@ -122,16 +132,29 @@ def parse_exprs_and_named_exprs(
             msg = f"Internal error: got output names {output_names}, but only got {len(native_series_list)} results"
             raise AssertionError(msg)
         native_results.update(zip(output_names, native_series_list))
+        returns_scalar.extend(
+            [expr._returns_scalar and expr._function_name != "lit"] * len(output_names)
+        )
     for col_alias, expr in named_exprs.items():
         native_series_list = expr._call(df)
         if len(native_series_list) != 1:  # pragma: no cover
             msg = "Named expressions must return a single column"
             raise ValueError(msg)
         native_results[col_alias] = native_series_list[0]
-    return native_results
+        returns_scalar.append(expr._returns_scalar and expr._function_name != "lit")
 
+    if all(returns_scalar) and not with_columns_context:
+        return native_results
+    else:
+        return {
+            col_name: col.over(Window.partitionBy(F.lit(1))) if _returns_scalar else col
+            for (col_name, col), _returns_scalar in zip(
+                native_results.items(), returns_scalar
+            )
+        }
```

Comment:
i'm not sure if this is the right place to do this

if we do `select` with all aggregations, we should just call `native_frame.agg`, which is what we do for DuckDB

Reply:
i think perhaps we should:
- only broadcast (which for pyspark means ".over(Window.partitionBy(F.lit(1))") if there's binary operation
- in `select`, call `native_frame.agg` if all expressions are aggregations
- in `group_by`, just pass the expressions though (like we do for duckdb)

Reply:
TIL about pyspark `agg`.

It makes sense, I will check a few edge cases and differences to take into account between `select` and `with_columns`. 
For group by, I think the logic is still a bit rudimental and we are not passing the expression itself. We might need to come back to that